README.md
<div align="center">
    بسم الله الرحمن الرحيم
</div>

# Good Loader

Display an elegant loader for your TALL Stack application.

### Description

When you load the site, and the page is already loaded, the package's loader view beautifully fades out in a specified `transition` time...

Otherwise, a customizable **`sentence`** loads along the view, as soon as your specified **`font`** is loaded; and until the page is ready; before everything fading all out in the end.

The result is something like this:

https://user-images.githubusercontent.com/81492351/179066243-16001c9c-330a-4ac7-a020-0e6d1adcc7c6.mp4


## Installation

This package is built for the [TALL stack](https://tallstack.dev). You must account for its requirements at first.

### Requirements

- [TailwindCSS](https://tailwindcss.com)
- [Alpine.js](https://alpinejs.dev)
- [Laravel](https://laravel.com)

### Steps

1. Install the package via Composer:

   ```bash
   composer require goodm4ven/good-loader
   ```

2. Publish the package configuration and view files:

   ```bash
   php artisan vendor:publish --provider="GoodM4ven\GoodLoader\GoodLoaderServiceProvider"
   ```

   - If you're **updating** the package, please add `--force` option to **override** the old assets.

3. Open the published `config/good-loader.php` file and specify your `font` name.

   - <details>
       <summary>
         Here are the <b>default</b> configurations of the file.
       </summary><br>

     ```php
     /*
      |--------------------------------------------------------------------------
      | Loading Font
      |--------------------------------------------------------------------------
      |
      | Provide the name of the font you're using, so that the sentence won't
      | load until that font is loaded at least.
      |
      */

     'font' => env('GOOD_LOADER_FONT', 'Mulish'),


     /*
      |--------------------------------------------------------------------------
      | Loading Sentence
      |--------------------------------------------------------------------------
      |
      | Customize the sentence which shows up at the center of the screen before
      | the page is completely loaded.
      |
      */

     'sentence' => env('GOOD_LOADER_SENTENCE', 'Loading...'),


     /*
      |--------------------------------------------------------------------------
      | Loading Transitions
      |--------------------------------------------------------------------------
      |
      | The time it takes to transition (fade) the `background` and the `sentence`.
      |
      */

     'transitions' => [
         'background' => env('GOOD_LOADER_TRANSITIONS_BACKGROUND', 1000),
         'sentence' => env('GOOD_LOADER_TRANSITIONS_SENTENCE', 300),
     ],


     /*
      |--------------------------------------------------------------------------
      | Loading Durations
      |--------------------------------------------------------------------------
      |
      | The time it takes for the `sentence` to start `animating`, as things are
      | still not loaded...
      |
      */

     'durations' => [
         'sentence-animating' => env('GOOD_LOADER_DURATIONS_SENTENCE_ANIMATING', 750),
     ],
     ```
     </details>

4. Ensure that your `content` in the `tailwind.config.js` file include the published view - along your font:

   ```js
   const defaultTheme = require('tailwindcss/defaultTheme');

   module.exports = {
       content: [
           ...
           './resources/views/**/*.blade.php', // or to './resources/views/vendor/good-loader/**' specifically...
       ],
       ...
       theme: {
           extend: {
               fontFamily: {
                   mulish: ['Mulish', ...defaultTheme.fontFamily.sans],
                   ...
   ```

   - The view is available at `resources/views/vendor/good-loader/partials/good-loader.blade.php` for customization.

5. Add the following Blade directive to your `master` layout or view - along the font:

   ```html
       ...
       <link rel="prefetch"
             href="https://fonts.googleapis.com/css2?family=Mulish:ital,wght@0,200;0,300;0,400;0,500;0,600;0,700;0,800;0,900;1,200;1,300;1,400;1,500;1,600;1,700;1,800;1,900">
   </head>
   <body class="font-mulish">
       @goodLoader
       ...
   ```

6. Finally, compile Alpine in `app.js` and the view's Tailwind classes in `app.css` with NPM:

   ```bash
   npm run dev
   ```


## Development

Feel free to **p**ull-**r**equest at any time. It's not like the package is moving anywhere... -Or is it? 😲

- There's an available and automated [CHANGELOG](CHANGELOG.md) of all the updates.

### TODOs

- Provide more time customizations.
- Neat background grids to pick from!
- Write [Cypress](https://cypress.io) and [Pest](https://pestphp.com/) tests! 🥲


## Remarks

Please do 🌟 all the packages you rely on, but NOT this one. 😍


<div align="center">
    <br>والحمد لله رب العالمين
</div>


CHANGELOG.md
# Changelog

All notable changes to `good-loader` will be documented in this file.

## Hotfix v1.1.1 - 2022-07-28

**Full Changelog**: https://github.com/GoodM4ven/good-loader/compare/v1.1.0...v1.1.1

## Release v1.1.0 - 2022-07-27

### What's Changed

- Update README.md by @GoodM4ven in https://github.com/GoodM4ven/good-loader/pull/1

### New Contributors

- @GoodM4ven made their first contribution in https://github.com/GoodM4ven/good-loader/pull/1

**Full Changelog**: https://github.com/GoodM4ven/good-loader/compare/v1.0.0...v1.1.0

## Release v1.0.0 - 2022-07-14

**Full Changelog**: https://github.com/GoodM4ven/good-loader/commits/v1.0.0


src/GoodLoaderServiceProvider.php
<?php

namespace GoodM4ven\GoodLoader;

use Illuminate\Support\Facades\Blade;
use Spatie\LaravelPackageTools\Package;
use Spatie\LaravelPackageTools\PackageServiceProvider;

class GoodLoaderServiceProvider extends PackageServiceProvider
{
    public function configurePackage(Package $package): void
    {
        /*
         *
         * This class is a Package Service Provider
         *
         * More info: https://github.com/spatie/laravel-package-tools
         *
         */
        $package
            ->name('good-loader')
            ->hasConfigFile()
            ->hasViews();
    }

    public function bootingPackage()
    {
        $this->registerBladeDirectives();
    }

    protected function registerBladeDirectives()
    {
        Blade::directive('goodLoader', function () {
            return "<?= view('good-loader::partials.good-loader')->render(); ?>";
        });
    }
}


resources/views/partials/good-loader.blade.php
<!-- Good Loader -->
<div class="fixed z-[99] grid h-full w-full items-center bg-white transition-opacity"
     style="transition-duration: {{ config('good-loader.transitions.background') }}ms;"
     x-data="{}"
     x-init="$nextTick(() => {
         clearTimeout(window.goodLoader);

         const $sentence = document.getElementById('good-loader-sentence');

         $sentence.classList.add('prevent-animating');
         $sentence.classList.add('opacity-0');
         $el.classList.add('opacity-0');
     });">
    <!-- Background -->
    <div id="good-loader-background"
         class="fixed z-[98] h-full w-full bg-white transition-opacity"
         style="transition-duration: {{ config('good-loader.transitions.sentence') }}ms;">
    </div>

    <!-- Sentence -->
    <div id="good-loader-sentence"
         class="text-center text-xl opacity-0 transition-opacity"
         style="transition-duration: {{ config('good-loader.transitions.sentence') }}ms;">
        {{ config('good-loader.sentence') }}
    </div>

    <!-- Script -->
    <script>
        setTimeout(() => {
            document.fonts.load('1.25rem "{{ config('good-loader.font') }}"').then(() => {
                window.goodLoader = setTimeout(() => {
                    if (document.readyState !== 'complete') {
                        const $background = document.getElementById('good-loader-background');
                        const $sentence = document.getElementById('good-loader-sentence');

                        $background.classList.add('opacity-0');

                        setTimeout(() => {
                            $sentence.classList.remove('opacity-0');
                        }, {{ config('good-loader.transitions.sentence') }});

                        setTimeout(() => {
                            if (!$sentence.classList.contains('prevent-animating')) {
                                $sentence.classList.add('animate-pulse');
                            }
                        }, {{ config('good-loader.durations.sentence-animating') }});
                    }
                }, 750);
            });
        }, 0);
    </script>
</div>


config/good-loader.php
<?php

return [

    /*
     |--------------------------------------------------------------------------
     | Loading Sentence
     |--------------------------------------------------------------------------
     |
     | Customize the sentence which shows up at the center of the screen before
     | JavaScript is completely loaded.
     |
     */

    'sentence' => env('GOOD_LOADER_SENTENCE', 'Loading...'),


    /*
     |--------------------------------------------------------------------------
     | Loading Font
     |--------------------------------------------------------------------------
     |
     | Provide the name of the font you're using, so that the sentence won't
     | load until that font is loaded at least; after CSS.
     |
     */

    'font' => env('GOOD_LOADER_FONT', 'Mulish'),


    /*
     |--------------------------------------------------------------------------
     | Loading Duration
     |--------------------------------------------------------------------------
     |
     | The minimum time it takes to keep the background whilst `loading`, even
     | after the Javascript is completely loaded.
     |
     | And the least duration it takes for the `sentence` to stay visible when
     | it gets displayed.
     |
     */

    'duration' => [
        'loading' => env('GOOD_LOADER_LOADING_DURATION', 500),
        'sentence' => env('GOOD_LOADER_SENTENCE_DURATION', 750),
    ],

];
